storage_broker/proto/broker.proto

```diff
@@ -48,3 +54,61 @@ message TenantTimelineId {
     bytes tenant_id = 1;
     bytes timeline_id = 2;
 }
+
+message FilterTenantTimelineId {
+    enum Mode {
+        // Invalid mode.
+        INVALID = 0;
+
+        // Subscribe to all timelines.
+        ALL_TIMELINES = 1;
+
+        // Subscribe to a specific timeline.
+        SPECIFIC_TIMELINE = 2;
+    }
+
+    Mode mode = 1;
+
+    // Valid only if mode == SPECIFIC_TIMELINE.
+    optional TenantTimelineId tenant_timeline_id = 2;
+}
+
+message SubscribeByFilterRequest {
+    // Subscription will emit messages only of the specified types.
+    repeated MessageType types = 1;
+
+    // Subscription will emit messages only for the specified tenant/timeline.
+    FilterTenantTimelineId tenant_timeline_id = 2;
+}
+
+enum MessageType {
+    UNKNOWN = 0;
+    SAFEKEEPER_TIMELINE_INFO = 2;
+    SAFEKEEPER_DISCOVERY_REQUEST = 3;
+    SAFEKEEPER_DISCOVERY_RESPONSE = 4;
+}
+
+// A message with a type.
+message TypedMessage {
+    MessageType type = 1;
+
+    optional SafekeeperTimelineInfo safekeeper_timeline_info = 2;
+    optional SafekeeperDiscoveryRequest safekeeper_discovery_request = 3;
+    optional SafekeeperDiscoveryResponse safekeeper_discovery_response = 4;
```

Comment:
What's the difference between SafekeeperTimelineInfo and SafekeeperDiscoveryResponse?

Reply:
SafekeeperDiscoveryResponse has less fields than SafekeeperTimelineInfo.
For example, it doesn't have `peer_horizon_lsn` and other fields, which are not necessary for establishing a connection.